### PR TITLE
feat: add get device by id, get devices and get history - GraphQL

### PR DESCRIPTION
**Pull request by juliamourac**
* **Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature.

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**
Adds queries to retrieve a device by id and a list of devices.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No.

* **Is there any issue related to this PR in other repository?** (such as dojot/dojot)
dojot/dojot#1388

* **Other information**:

### CODE REVIEW

File: src/graphql/device/Resolvers.js [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+            device.attrs.push({
+              label: deviceData.attrs[key][i].label,
+              type: valueType,
+            });
+          }
+        });
+
+        return (device);
+      } catch (err) {
+        LOG.warn(err);
```
Comment (mprevide) [inline src/graphql/device/Resolvers.js:39]: Log Error :)

File: src/graphql/device/Resolvers.js [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+      try {
+        const { data: fetchedData } = await axios(optionsAxios(UTIL.GET, requestString));
+        const deviceList = ([{
+          totalCount: fetchedData.devices.length,
+          hasNextPage: fetchedData.pagination.has_next,
+          hasPreviousPage: fetchedData.pagination.page != 1,
+          devices: fetchedData.devices,
+        }]);
+
+        console.log(deviceList);
```
Comment (mprevide) [inline src/graphql/device/Resolvers.js:65]: Logger

File: src/graphql/device/Resolvers.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+        return deviceList;
+      } catch (error) {
+        LOG.warn(error);
```
Comment (mprevide) [inline src/graphql/device/Resolvers.js:69]: Error

File: src/graphql/device/Resolvers.js [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+      }
+    },
+
+    getDeviceHistory(root, params){
+      const history = [];
+      const keys = Object.keys(params.input);
+      keys.shift();
+      console.log(`Keys: ${keys}`);
```
Comment (mprevide) [inline src/graphql/device/Resolvers.js:77]: Remove

File: src/graphql/device/TypeDefs.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
+const TypeDefs = [`
+    enum Type{
+        dynamic
+        static
+        actuate
+    }
+    enum ValueType {
+        integer
```
Comment (mprevide) [inline src/graphql/device/TypeDefs.js:8]: Enums should be uppercase?

File: src/graphql/user/Resolvers.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
+const axios = require('axios');
+const UTIL = require('../utils/AxiosUtils');
+const LOG = require('../../utils/Log');
+const connection = require('../../db');
+
+const Resolvers = {
```
Comment (mprevide) [inline src/graphql/user/Resolvers.js:6]: It would be better if the user / persistence feature was not in this PR. :)

**Comment by mprevide** [comment]
Why codecov report doesn't run?

File: src/graphql/device/TypeDefs.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
+const TypeDefs = [`
+    enum Type{
```
Comment (mprevide) [inline src/graphql/device/TypeDefs.js:2]: Remove Type

File: src/graphql/device/TypeDefs.js (hunk 5 above)
Comment (mprevide) [reply src/graphql/device/TypeDefs.js:8]: ``
enum ValueType {
   NUMBER
   STRING
   BOOLEAN
   GEO
   UNDEFINED
}
` `

File: src/graphql/device/TypeDefs.js [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        float
+        text
+        number
+        string
+        boolean
+        geo
+        undefined
+    }
+    type Meta{
+        label: String!
```
Comment (mprevide) [inline src/graphql/device/TypeDefs.js:18]: Remove Meta

File: src/graphql/device/TypeDefs.js [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        desc: String
+        valueType: String!
+        value: String!
+    }
+    type Attr{
```
Comment (mprevide) [inline src/graphql/device/TypeDefs.js:23]: type Attr {
   label: String!
   type: ValueType!
}

File: src/graphql/device/TypeDefs.js [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        id: String!
+        #Device label
+        label: String!
+        #Attributes
+        attrs: [Attr!]
+        #Template
+        templates: [Template!]!
+    }
+    #Paging things
+    type DeviceListPage {
```
Comment (mprevide) [inline src/graphql/device/TypeDefs.js:42]: # A paginated list of Devices.
type DeviceListPage {
   # Identifies the total count of item.
   totalPages: Int!
   # Indicate the current Page.
   currentPage: Int!
   # List of Devices.
   devices: [Device]
}

File: src/graphql/Schema.js [hunk 11]
```diff
@@ -3,21 +3,33 @@ const { merge } = require('lodash');
...
-    #Checks if templates has Image Firmware and return a array with objects key-value, where key is a id template and value is a boolean.
-    #The value is true if the template has image firmware.
-    templatesHasImageFirmware(templatesId: [Int]!): [MapStringToString]
+  type Query {
+    #Get a template by Id
+      template(id: Int!): Template
+      #Checks if templates has Image Firmware and return a array with objects key-value, where key is a id template and value is a boolean.
+      #The value is true if the template has image firmware.
+      templatesHasImageFirmware(templatesId: [Int]!): [MapStringToString]
+      getDevices(pagesize: Int, pagenumber: Int, label: String): [DeviceListPage]
```
Comment (mprevide) [inline src/graphql/Schema.js:19]: getDevices(page: PageInput filter: FilterDeviceInput): [DeviceListPage]

File: src/graphql/Schema.js [hunk 12]
```diff
@@ -3,21 +3,33 @@ const { merge } = require('lodash');
...
+      getDeviceById(deviceId: String): Device
```
Comment (mprevide) [inline src/graphql/Schema.js:20]: getDeviceById(deviceId: String!): Device

File: src/graphql/Schema.js [hunk 13]
```diff
@@ -3,21 +3,33 @@ const { merge } = require('lodash');
...
+      getDeviceHistory(input: HistoryInput): [History]
```
Comment (mprevide) [inline src/graphql/Schema.js:21]: getHistory(filter: HistoryInput!): [History]

File: package.json [hunk 14]
```diff
@@ -4,7 +4,7 @@
   "description": "This repository is to get together all services can help GUI",
   "main": "index.js",
   "scripts": {
-    "test": "jest --coverage && codecov",
+    "test": "jest src/tests/**/*.js --coverage && codecov",
```
Comment (mprevide) [inline package.json:7]: Change to src/tests/.*.test.js

**Comment by mprevide** [comment]
When Travis run the test:

``
 PASS  src/tests/graphql/resolvers/Device.test.js
[2020-04-22T13:34:37.547] [WARN] BackStage - TypeError: Cannot read property 'devices' of undefined
    at Object.getDevices (/home/travis/build/dojot/backstage/src/graphql/device/Resolvers.js:1724:35)
    at <anonymous>
``

**Comment by mprevide** [comment]
No need to worry about the total percentage of codecov, just cover the new device, history and config files

File: src/graphql/device/Resolvers.js [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+};
+const setToken = ((token) => {
+  params.token = token;
+});
+const optionsAxios = ((method, url) => UTIL.optionsAxios(method, url, params.token));
+
+const Resolvers = {
+  Query: {
+
+    async getDeviceById(root, { deviceId }, context) {
```
Comment (mprevide) [inline src/graphql/device/Resolvers.js:16]: It shoud ignore attrs != dinamyc

File: src/graphql/device/Resolvers.js [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+        });
+
+        return (device);
+      } catch (err) {
+        LOG.error(err);
+      }
+    },
+
+    async getDevices(root, params) {
+      // building the request string
```
Comment (mprevide) [inline src/graphql/device/Resolvers.js:44]: It shoud ignore attrs != dinamyc

File: src/graphql/device/Resolvers.js [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+      keys.shift();
+      const requestStringPt1 = '/history/device/';
+      let requestStringPt2 = '/history';
+      
+      if (keys.length != 0) {
+        requestStringPt2 += '?';
+        const lastKey = keys[keys.length - 1];
+        keys.forEach((element) => {
+          if (element === lastKey) {
+            requestStringPt2 += `${element}=${params[element]}`;
```
Comment (mprevide) [inline src/graphql/device/Resolvers.js:91]: Note: When the history api receives more than one attr, the history has a different return.

Comment (mprevide) [reply src/graphql/device/Resolvers.js:91]: Note 2: When there is still no history for that attr and only the history of an attr is requested, the api returns 404.

File: src/graphql/device/TypeDefs.js [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+ }
+  
+ input FilterDeviceInput {
+    # Return only devices that are named accordingly
+    label: String # (prefix or suffix match)
+ }
+  
+ input HistoryInput {
+    devices: [HistoryDeviceInput]!
+    beginTime: String # unix time,usar scalar DateTime?? será optional qdo tem lastN ??
```
Comment (mprevide) [inline src/graphql/device/TypeDefs.js]: Remove these comments in pt and maybe add others in english

File: src/graphql/Schema.js [hunk 19]
```diff
@@ -3,21 +3,31 @@ const { merge } = require('lodash');
...
-    templatesHasImageFirmware(templatesId: [Int]!): [MapStringToString]
+  type Query {
+    #Get a template by Id
+      template(id: Int!): Template
+      #Checks if templates has Image Firmware and return a array with objects key-value, where key is a id template and value is a boolean.
+      #The value is true if the template has image firmware.
+      templatesHasImageFirmware(templatesId: [Int]!): [MapStringToString]
+      getDevices(page: PageInput, filter: FilterDeviceInput): [DeviceListPage]
+      getDeviceById(deviceId: String!): Device
+      getDeviceHistory(input: HistoryInput): [History]
```
Comment (mprevide) [inline src/graphql/Schema.js]: Why not filter instead of input?

File: src/graphql/Schema.js [hunk 20]
```diff
@@ -3,21 +3,31 @@ const { merge } = require('lodash');
...
+      getConfig(tenant: String!, user:String!): Config
+    }
+  type Mutation {
```
Comment (mprevide) [inline src/graphql/Schema.js:22]: I think It's better remove everything related with  user and config in this PR. :)

File: src/tests/graphql/resolvers/Device.test.js [hunk 21]
```diff
@@ -0,0 +1,147 @@
+const axios = require('axios');
```
Comment (mprevide) [inline src/tests/graphql/resolvers/Device.test.js:1]: 85% ou more

File: src/graphql/Schema.js [hunk 22]
```diff
@@ -3,21 +3,31 @@ const { merge } = require('lodash');
...
-    #Checks if templates has Image Firmware and return a array with objects key-value, where key is a id template and value is a boolean.
-    #The value is true if the template has image firmware.
-    templatesHasImageFirmware(templatesId: [Int]!): [MapStringToString]
+  type Query {
+      #Get a template by Id
+      template(id: Int!): Template
+      #Checks if templates has Image Firmware and return a array with objects key-value, where key is a id template and value is a boolean.
+      #The value is true if the template has image firmware.
+      templatesHasImageFirmware(templatesId: [Int]!): [MapStringToString]
+      getDevices(page: PageInput, filter: FilterDeviceInput): [DeviceListPage]
```
Comment (mprevide) [inline src/graphql/Schema.js:17]: Add comments before each query or mutation for used in documentation

File: src/graphql/device/TypeDefs.js [hunk 23]
```diff
@@ -0,0 +1,75 @@
+const TypeDefs = [`
+
+type Device {
```
Comment (mprevide) [inline src/graphql/device/TypeDefs.js:3]: Add comments before each type, etc for to be used in documentation

File: src/graphql/device/Resolvers.js [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+            return Promise.resolve(null);
+          });
+          historyPromiseArray.push(promiseHistory);
+        });
+        const promiseDevice = axios(optionsAxios(UTIL.GET, `/device/${obj.deviceID}`));
+        devicePromiseArray.push(promiseDevice);
+      });
+
+      //API calls are made and results are saved in arrays
+      await Promise.all(historyPromiseArray).then(values => {
```
Comment (mprevide) [inline src/graphql/device/Resolvers.js:121]: Is everything working fine when are you request more then one and one  of then without history now, when return  just one 404?

File: src/graphql/device/Resolvers.js [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+
+            device.attrs.push({
+              label: deviceData.attrs[key][i].label,
+              type: valueType,
+            });
+          }
+        });
+
+        return (device);
+      } catch (err) {
```
[the remark this answers is not in the document]
Comment (juliamourac) [reply src/graphql/device/Resolvers.js:37]: If using the graphql interface, the following object is returned:
{
  "data": {
    "getDeviceById": null
  }
}

While at the console the failed request is printed.

File: src/graphql/device/Resolvers.js [hunk 26]
```diff
@@ -0,0 +1,218 @@
...
+    },
+
+    async getDeviceHistory(root, params) {
+      let history = [];
+      const keys = Object.keys(params.filter);
+      keys.shift();
+      const requestStringPt1 = '/history/device/';
+      let requestStringPt2 = '/history';
+
+      if (keys.length != 0) {
```
[the remark this answers is not in the document]
Comment (juliamourac) [reply src/graphql/device/Resolvers.js:95]: Could I just change the type def (beginTime, endTime to dateFrom, dateTo)? So I don't need to add verifications if these parameters were sent or not.

File: src/graphql/device/Resolvers.js (hunk 24 above)
Comment (juliamourac) [reply src/graphql/device/Resolvers.js:121]: It is, if there's entries for the other device, it shows only the information about it and ignores entries for the missing one. (The last unit test added is about this scenario)

**Comment by mprevide** [comment]
Just fix codefactor and deepscan issues :)

File: src/tests/graphql/resolvers/Device.test.js [hunk 27]
```diff
@@ -0,0 +1,869 @@
...
+    expect(output).toEqual([
+      {
+        "totalPages": 1,
+        "currentPage": 1,
+        "devices": [
+          {
+            "attrs": [
+              {
+                "label": "attr1",
+                "meta": null,
```
Comment (mprevide) [inline src/tests/graphql/resolvers/Device.test.js:111]: Meta and type should be here?

File: src/tests/graphql/resolvers/Device.test.js [hunk 28]
```diff
@@ -0,0 +1,869 @@
...
+                  "meta": null,
+                  "type": "dynamic",
+                  "value": null,
+                  "value_type": "NUMBER"
+                }
+              ]
+            },
+            "id": "65f9d7",
+            "label": "disp10",
+            "templates": [
```
Comment (mprevide) [inline src/tests/graphql/resolvers/Device.test.js:161]: Templates?

File: src/tests/graphql/resolvers/Device.test.js [hunk 29]
```diff
@@ -0,0 +1,813 @@
...
+          {
+            created: '2020-03-09T17:10:34.364406+00:00',
+            id: 1,
+            is_static_overridden: false,
+            label: 'temperature',
+            metadata: [
+              {
+                created: '2020-03-09T17:10:34.369905+00:00',
+                id: 2,
+                'is_static´_overridden': false,
```
Comment (mprevide) [inline src/tests/graphql/resolvers/Device.test.js:344]: `is_static´_overridden` should be `is_static_overridden`

File: src/graphql/device/TypeDefs.js [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+    NUMBER
+    STRING
+    BOOLEAN
+    GEO
+    UNDEFINED
+ }
+ """Necessary data about and attribute of a device"""
+ type Attr {
+    label: String!
+    value_type: ValueType!
```
Comment (mprevide) [inline src/graphql/device/TypeDefs.js:19]: change to valueType

File: src/tests/graphql/resolvers/Device.test.js [hunk 31]
```diff
@@ -0,0 +1,813 @@
...
+    {
+      label: 'temperature',
+      valueType: 'NUMBER',
+      value: 36.5,
+      timestamp: '2018-03-22T13:46:21.535000Z',
+    }],
+  }]);
+});
+
+/*it ('should return formatted device information', () => {
```
Comment (mprevide) [inline src/tests/graphql/resolvers/Device.test.js:761]: Just remove this commented code :)

**Pull request merged**